### PR TITLE
Packaging as standalone NWJS app

### PR DESCRIPTION
**Pull request by Sjlver**
This commit introduces a multi-directory app structure:

```
app
├── app.html     -- Main HTML file
├── app.js       -- Background script (for Chrome app use only)
├── images       -- Images used directly by the app
├── scripts      -- Our own JavaScript code goes here
├── stylesheets
└── vendor       -- Third-party code goes here

build            -- The built app ends up here (after transpiling etc.)
node_modules     -- These are all development dependencies (e.g., gulp).
                    Those files are not accessible to the app.
releases         -- Installers, packages and dmg files end up here
resources        -- Resource files like logos for the installer
tasks            -- Gulp tasks for building, creating releases, ...
test             -- Unit tests
```

We can now package the app. This works under Mac OS. I've also added
untested code for Windows and Linux. Run `npm run release` on the
corresponding operating system to create a build.

In addition, there are a number of little fixes. We can now run multiple
test suites without conflicts. There is an editorconfig and jshintrc
file, ...

I renamed the application from onlykey-chrome-app to onlykey-app, now
that it also works in other runtimes than Chrome.

Much of the code for building/releasing is based on
<https://github.com/szwacz/nw-boilerplate>, with thanks.

This will no longer run as Chrome app as is, but it does still work. Try `npm run chrome`.

A first release for mac os is here: <https://drive.google.com/open?id=0B8pYrLJnKwDSQUx0RFhZTXBzdWM>

### CODE REVIEW

**Comment by onlykey** [comment]
@Sjlver You are the man! I just tested on Debian Linux. 

Step 1 - Install Node
$ curl -sL https://deb.nodesource.com/setup_6.x | sudo -E bash -
$ sudo apt-get install -y nodejs
Step 2 - Install modules 
$ cd <repo path>
$ npm install gulp gulp-sourcemaps gulp-less fs-jetpack q nw
$ npm run release
...
Creating DEB package...
[11:46:31] DEB package ready! /app/releases/onlykey-app_4.0.1_amd64.deb
$ dpkg -i releases/onlykey-app_4.0.1_amd64.deb 

Launch app
Success!

**Comment by TheSirC** [comment]
Sorry I tried on Windows and it's not working 😢 
Here is the relevant  part of `npm-debug.log` : 
```
2 info using npm@3.10.10
3 info using node@v6.11.3
...
\OnlyKey-Chrome-App
11 silly lifecycle onlykey-app@4.0.1~release: Args: [ '/d /s /c', 'gulp release --env=production' ]
12 silly lifecycle onlykey-app@4.0.1~release: Returned: code: 1  signal: null
13 info lifecycle onlykey-app@4.0.1~release: Failed to exec release script
14 verbose stack Error: onlykey-app@4.0.1 release: `gulp release --env=production`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\lib\utils\lifecycle.js:255:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\lib\utils\spawn.js:40:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:920:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:230:5)
15 verbose pkgid onlykey-app@4.0.1
```

**Comment by Sjlver** [comment]
@TheSirC Thanks for trying. From the error message, it's not clear to me what the error ist... apparently executing `gulp release` fails. Just to be sure... did you run `npm install` before running `npm run release`? That is required to install the gulp dependency.

Any chance you could get me the error message printed by gulp? Or try running `node_modules/.bin/gulp release --env=production` manually.

**Comment by TheSirC** [comment]
@Sjlver Thanks for your amazing work and your time ! 
I did run `npm install` before running `npm run release`.
Here is the error printed by gulp : 
```
> gulp release --env=production

[11:23:51] Using gulpfile G:\Claude-Alban\Documents\Personnel\Projets\OnlyKey-Chrome-App\gulpfile.js
[11:23:51] Starting 'clean'...
[11:23:51] Finished 'clean' after 11 ms
[11:23:51] Starting 'transpile'...
[11:23:51] Starting 'copy'...
[11:23:51] Starting 'finalize'...
[11:23:51] Finished 'finalize' after 4.08 ms
[11:23:51] Finished 'transpile' after 126 ms
[11:24:00] Finished 'copy' after 8.16 s
[11:24:00] Starting 'build'...
[11:24:00] Finished 'build' after 6.28 μs
[11:24:00] Starting 'release'...
[11:24:54] Building installer with NSIS...
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: spawn makensis ENOENT
    at exports._errnoException (util.js:1020:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:197:32)
    at onErrorNT (internal/child_process.js:376:16)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)

npm ERR! Windows_NT 10.0.15063
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "release"
npm ERR! node v6.11.3
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! onlykey-app@4.0.1 release: `gulp release --env=production`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the onlykey-app@4.0.1 release script 'gulp release --env=production'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the onlykey-app package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     gulp release --env=production
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs onlykey-app
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls onlykey-app
npm ERR! There is likely additional logging output above.
```

**Comment by Sjlver** [comment]
@TheSirC Gotta look into this. Sorry I don't have easy access to a Windows machine, but will borrow one soon.

My hunch is that one has to install NSIS separately. The code is based on nwjs-boilerplate, and they mention something like that here: https://github.com/szwacz/nw-boilerplate#special-precautions-for-windows

Sorry, I should have added that to the README...

**Comment by onlykey** [comment]
I got some time to test out on Windows and I am stuck the same place as @TheSirC 
I made sure to install NSIS and make sure it was in the recommended path but no luck. I tried messing around with dependencies as mentioned here - https://github.com/szwacz/nw-boilerplate/issues/21 but no luck. Let me know if you have any ideas I can try @Sjlver

**Comment by onlykey** [comment]
@Sjlver It looks like the nw.js boilerplate is 2 years old and may no longer be supported. Maybe there is a later boilerplate or example app that would work better - https://github.com/nw-cn/awesome-nwjs

**Comment by onlykey** [comment]
@Sjlver @TheSirC OK I got it working. I didn't reopen the cmd prompt after installing NSIS and manually setting the environment path. After reopening the cmd prompt I was able to run 'npm run release' with success. Fyi, I don't think it matters but I did change my dependencies in package.json to:

  "devDependencies": {
    "fs-jetpack": "^1.2.0",
	"gulp": "^3.8.8",
	"gulp-babel": "^6.1.2",
	"babel-preset-es2015": "^6.6.0",
    "gulp-less": "^3.3.2",
    "gulp-sourcemaps": "^2.6.1",
	"gulp-util": "^3.0.1",
    "nw": "^0.24.4",
    "q": "^1.0.1",
    "tree-kill": "^1.1.0",
    "yargs": "^8.0.2"
	},
	"optionalDependencies": {
	"appdmg": "^0.3.6"
	},

Here is the compiled release file for windows - https://drive.google.com/file/d/0B7SBNa10_4RhNlFwVWJvQmRvMzQ/view?usp=sharing

It's an installer file and some observations:

- Tested on Windows 7
- On install AV inspects it but decides its not a threat and installs
- It has an NWBP icon - 
image

Other than that the app appears to be completely functional.

This is great work guys, we now have standalone app working on OSX, Linux, and Windows! 

Do you think we should stick with this boilerplate and NSIS or go another route? It looks like there are some other options for boilerplate and example NW.js apps - https://github.com/naviapps/create-nw-react-app

**Comment by TheSirC** [comment]
I'm going to test it on Windows 10 as soon as possible. @onlykey I was considering creating a Elctron app for it but I would really love to have either a hand or a "mentor"

**Comment by onlykey** [comment]
@TheSirC Electron apps are more popular. The benefit of NW.js is that it supports Chrome APIs. Since the original app is a Chrome app this allows us to do a native app with little changes. This should also allow us to just maintain one app. The Chrome app will continue to be supported on Chromebook and the native app will be supported on Mac, Linux, Windows and they will share the same js and html files so when we add new features we only have to do it once rather than twice.

**Comment by onlykey** [comment]
@Sjlver I tried 'npm run chrome' to run as chrome app. This works! I am impressed you did a really good job with this. I was able to load the build directory as an unpacked extension and it works as normal. 
image

@rodgolpe We have tested this app on Windows, Linux, Mac, and as a Chrome App. Can you give it a final review and then we can merge this?

File: config/env_production.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "production",
```
Comment (rodgolpe) [inline config/env_production.json]: Is this `name` collision with that defined in `env_chrome.json` any cause for concern?

File: test/configure-slot-test.js [hunk 2]
```diff
@@ -1,57 +1,36 @@
...
-            .build();
-    });
-
-    after(function() {
-        //return driver.quit();
-    });
-
     function expectDialogOpen(id) {
-        var dialog = driver.findElement(By.id(id));
+        let dialog = driver.findElement(By.id(id));
```
Comment (rodgolpe) [inline test/configure-slot-test.js]: Harmless question: why not `const` in all these tests? When would you want to allow `dialog` to be reassigned?

**Comment by rodgolpe** [comment]
@Sjlver When I run `npm test` it complains `The specified executable path does not exist: /Users/rgolpe/Repos/OnlyKey/OnlyKey-Chrome-App/node_modules/nw/nwjs/chromedriver`. What should I do besides `npm install`?

**Comment by Sjlver** [comment]
(Sorry for the late/short replies, I'm currently offline with only
occasional email access. Will try to catch up on the weekend)

@rodgolpe This is a special chromedriver that comes with NWJS. You probably
have to install the "SDK" flavor of NWJS for this to get installed. Try
`npm install nw --nwjs_build_type=sdk` and see [here](
https://github.com/trustcrypto/OnlyKey-Chrome-App/pull/69/commits/6b5dc18787b920f9050b87f253e5666c3c3b62a4)
for why this isn't the default.

@onlykey The logo for Windows is at `resources/windows/icon.ico`. I can
replace it on the weekend.

@rodgolpe Oops `env_chrome.json` should say that the name is "chrome". I
missed that during copy/paste. My bad.

@rodgolpe regarding `const` vs `let`, I guess it's a style question that we
could resolve either way. I'm happy to use `const` where possible. I think
it would be less readable to write `const DIALOG = ...`, but const with
camelCase should be fine.

That's it for the moment; more on the weekend, when I'm back.

**Comment by rodgolpe** [comment]
Hi @Sjlver -- did you have anymore commits for this PR? I thought there'd be at least one for some cleanup items. If not, I'll take a closer look this weekend and get it merged. Thanks!

**Comment by Sjlver** [comment]
Sorry... it's all on my Todo-list but got delayed. Looks like I'll have time this weekend, too.

**Comment by Sjlver** [comment]
Alright, finally got to doing this. Sorry for the delay.

I think this PR is now ready for merging.

**Pull request merged**